include/switch-language.php

```diff
@@ -0,0 +1,173 @@
+<?php
+/**
+ * @package Polylang
+ */
+
+/**
+ * Class to handle site language switch.
+ */
+class PLL_Switch_Language {
+
+	/**
+	 * @var PLL_Model
+	 */
+	private static $model;
+
+	/**
+	 * The previous language.
+	 *
+	 * @var PLL_Language|null
+	 */
+	public static $previous_language;
+
+	/**
+	 * The original language.
+	 *
+	 * @var PLL_Language|null
+	 */
+	private static $original_language;
+
+	/**
+	 * The current language.
+	 *
+	 * @var PLL_Language|null
+	 */
+	private static $current_language;
+
+	/**
+	 * Setups filters.
+	 *
+	 * @since 3.7
+	 *
+	 * @param PLL_Model $model Instance of `PLL_Model`.
+	 * @return void
+	 */
+	public static function init( PLL_Model $model ) {
+		self::$model = $model;
+
+		add_action( 'pll_language_defined', array( static::class, 'set_current_language' ) );
```

Comment:
I think we should do this early.

```suggestion
		add_action( 'pll_language_defined', array( static::class, 'set_current_language' ), -1000 );
```